src/NetSapiensSharp/Connector.cs

```diff
@@ -82,17 +95,40 @@ public Request<Tresponse> CreateRequest<Tresponse>(string action_path) where Tre
         public IRestResponse<Tresponse> Send<Tresponse>(Request<Tresponse> request)
             where Tresponse : class, new()
         {
-            Authenticate();
-            request._Request.AddHeader("Authorization", "Bearer " + _SessionToken);
-            var r = request._Client.Execute<Tresponse>(request._Request);
-            if (r.StatusCode == System.Net.HttpStatusCode.OK && r.Data == null && typeof(Tresponse) == typeof(Common.OK))
+            if (IsAccessTokenExpired())
+            {
+                _SessionToken = null;
+            }
+            IRestResponse<Tresponse> myResponse = null;
+            int myRetryCount = 1;
+
+            do
+            {
+                Authenticate();
+                request._Request.AddOrUpdateParameter("Authorization", "Bearer " + _SessionToken, ParameterType.HttpHeader);
+                myResponse = request._Client.Execute<Tresponse>(request._Request);
+
+                if (myResponse.StatusCode == System.Net.HttpStatusCode.Unauthorized)
+                {
+                    _SessionToken = null;
+                    myRetryCount++;
+                }
+                else if (myResponse.StatusCode == System.Net.HttpStatusCode.OK && myResponse.Data == null && typeof(Tresponse) == typeof(Common.OK))
+                {
+                    myResponse.Data = new Tresponse();
+                    myResponse.ErrorException = null;
+                    myResponse.ErrorMessage = null;
+                    myResponse.ResponseStatus = ResponseStatus.Completed;
+                    break;
+                }
+            }
+            while (_SessionToken == null && myRetryCount <= _UnauthorizedRetryLimit);
+
+            if (myRetryCount == _UnauthorizedRetryLimit)
             {
-                r.Data = new Tresponse();
-                r.ErrorException = null;
-                r.ErrorMessage = null;
-                r.ResponseStatus = ResponseStatus.Completed;
+                throw new Exception($"Failed to authorize. Retry limit of {_UnauthorizedRetryLimit} has been reached.");
```

Comment:
Instead of an exception, we should send back an error message / Tresponse. Exceptions should be reserved for when we cannot send back any kind of response that the user can action.

Reply:
@jbparker I had an idea I had attempted to implement but had issues earlier. I wanted to use `UnauthorizedAccessException` for the ErrorException but then I got an error
![image](https://user-images.githubusercontent.com/93213727/139707449-e117061c-6e7f-4c5f-ad3b-1052031b77ac.png)

Do we want to set the `ErrorException`?
`myResponse.Data = null;
  myResponse.ErrorException = UnauthorizedAccessException;
  myResponse.ErrorMessage = $"Failed to authorize. Retry limit of {_UnauthorizedRetryLimit} has been reached.";
  myResponse.ResponseStatus = ResponseStatus.Aborted;`

Reply:
I don't think that we have an exception to use, so you'll likely just have to leave this null unless the exception coming back from the response is already populated. But we wouldn't want to just create one here.

Reply:
@jbparker I just left it as null for now. The `ErrorMessage` and `ResponseStatus` can be checked if needed because those are populated.